test/src/test/java/jenkins/security/ResourceDomainTest.java

```diff
@@ -0,0 +1,281 @@
+package jenkins.security;
+
+import com.gargoylesoftware.htmlunit.Page;
+import com.gargoylesoftware.htmlunit.html.HtmlPage;
+import hudson.ExtensionList;
+import hudson.model.DirectoryBrowserSupport;
+import hudson.model.FreeStyleProject;
+import hudson.model.Item;
+import jenkins.model.Jenkins;
+import jenkins.model.JenkinsLocationConfiguration;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.jvnet.hudson.test.CreateFileBuilder;
+import org.jvnet.hudson.test.For;
+import org.jvnet.hudson.test.Issue;
+import org.jvnet.hudson.test.JenkinsRule;
+import org.jvnet.hudson.test.MockAuthorizationStrategy;
+
+import java.net.URL;
+import java.util.UUID;
+
+@Issue("JENKINS-41891")
+@For({ ResourceDomainRootAction.class, ResourceDomainFilter.class, ResourceDomainConfiguration.class })
+public class ResourceDomainTest {
+
+    @Rule
+    public JenkinsRule j = new JenkinsRule();
+
+    private static final String RESOURCE_DOMAIN = "127.0.0.1";
+
+    @Before
+    public void prepare() throws Exception {
+        String resourceRoot;
+        URL root = j.getURL(); // which always will use "localhost", see JenkinsRule#getURL()
+
+        resourceRoot = root.toString().replace("localhost", RESOURCE_DOMAIN);
+        ResourceDomainConfiguration configuration = ExtensionList.lookupSingleton(ResourceDomainConfiguration.class);
+        configuration.setUrl(resourceRoot);
+    }
+
+    @Test
+    public void secondDomainBasics() throws Exception {
+        JenkinsRule.WebClient webClient = j.createWebClient();
+
+        { // DBS directory listing is shown as always
+            Page page = webClient.goTo("userContent");
+            Assert.assertEquals("successful request", 200, page.getWebResponse().getStatusCode());
+            Assert.assertTrue("still on the original URL", page.getUrl().toString().contains("/userContent"));
+            Assert.assertTrue("web page", page.isHtmlPage());
+            Assert.assertTrue("complex web page", page.getWebResponse().getContentAsString().contains("javascript"));
```

Comment:
BTW does HtmlUnit implement any CSP protections—can it be used to verify that stuff really works?

Reply:
Unsure, it didn't really back in https://github.com/jenkinsci/javadoc-plugin/pull/4#issuecomment-169152175 but we know what would trigger and check for that -- the headers.

This one basically just asserts that we get `dir.jelly` instead of `plaindir.jelly`, by the way.